File: app/src/main/java/com/study/lc/studykotlin/greeny/Bennyhuo/Kt10/src/main/java/net/println/kt10/java/LazyThreadSafeSynchronized.java
package com.study.lc.studykotlin.greeny.Bennyhuo.Kt10.src.main.java.net.println.kt10.java;

/**
 * Created by benny on 11/13/16.
 */
public class LazyThreadSafeSynchronized {
    private static LazyThreadSafeSynchronized INSTANCE;

    private LazyThreadSafeSynchronized(){}

    public static synchronized LazyThreadSafeSynchronized getInstance(){
        if(INSTANCE == null){
            INSTANCE = new LazyThreadSafeSynchronized();
        }
        return INSTANCE;
    }
}

File: app/src/main/java/com/study/lc/studykotlin/greeny/Bennyhuo/Kt10/src/main/kotlin/net/println/kt10/kotlin/LazyThreadSafeSynchronized.kt
package net.println.kt10.kotlin

/**
 * Created by benny on 11/13/16.
 */
class LazyThreadSafeSynchronized private constructor() {
    companion object {
       private var instance: LazyThreadSafeSynchronized? = null//?可以为空

        @Synchronized
        fun get(): LazyThreadSafeSynchronized{
            if(instance == null) instance = LazyThreadSafeSynchronized()
            return instance!!//!!不可以为空
        }
    }

}